src/oracle/oracle.cairo

```diff
@@ -329,21 +397,301 @@ mod Oracle {
         /// * `data_store` - The data store.
         /// * `event_emitter` - The event emitter.
         /// * `params` - The set price params.
-        fn set_primary_prices(
+        fn set_prices_(
             ref self: ContractState,
             data_store: IDataStoreDispatcher,
             event_emitter: IEventEmitterDispatcher,
             params: SetPricesParams,
-        ) { // TODO
+        ) {
+            let validated_prices = self.validate_prices(data_store, params);
+
+            let mut length = validated_prices.len();
+            loop {
+                if length == 0 {
+                    break;
+                }
+                match validated_prices.get(length - 1) {
+                    Option::Some(i) => {
+                        let validated_price = *i.unbox();
+                        self
+                            .emit_oracle_price_updated(
+                                event_emitter,
+                                validated_price.token,
+                                validated_price.min,
+                                validated_price.max,
+                                false
+                            );
+                        self
+                            .set_primary_price_(
+                                validated_price.token,
+                                Price { min: validated_price.min, max: validated_price.max }
+                            );
+                    },
+                    Option::None(_) => {
+                        OracleError::DUPLICATED_TOKEN_PRICE();
+                    }
+                }
+                length -= 1;
+            }
         }
 
         /// Validate prices in params.
         /// # Arguments
         /// * `data_store` - The data store.
         /// * `params` - The set price params.
-        fn validate_prices(
+        fn validate_prices_(
             self: @ContractState, data_store: IDataStoreDispatcher, params: SetPricesParams,
-        ) { // TODO
+        ) -> Array<ValidatedPrice> {
+            let signers = self.get_signers_(data_store, @params);
+
+            let mut cache = oracle_utils::empty_set_prices_cache();
+
+            cache
+                .min_block_confirmations = data_store
+                .get_u128(keys::min_oracle_block_confirmations())
+                .try_into()
+                .unwrap();
+
+            cache
+                .max_price_age = data_store
+                .get_u128(keys::max_oracle_price_age())
+                .try_into()
+                .unwrap();
+
+            cache
+                .max_ref_price_deviation_factor = data_store
+                .get_u128(keys::max_oracle_ref_price_deviation_factor());
+
+            let mut signers_len = signers.len();
```

Comment:
shouldn't you loop on params.tokens first ?

Reply:
Then have two successive nested loops on signers

Reply:
not sure what you mean, in the solidity `cache.min_block_confirmation` `cache.max_price_age ` and ` `cache.max_ref_price_deviation_factor` are assigned before the iteration:
```
SetPricesCache memory cache;

        cache.validatedPrices = new ValidatedPrice[](params.tokens.length);
        cache.minBlockConfirmations = dataStore.getUint(Keys.MIN_ORACLE_BLOCK_CONFIRMATIONS);
        cache.maxPriceAge = dataStore.getUint(Keys.MAX_ORACLE_PRICE_AGE);
        cache.maxRefPriceDeviationFactor = dataStore.getUint(Keys.MAX_ORACLE_REF_PRICE_DEVIATION_FACTOR);

        for (uint256 i; i < params.tokens.length; i++) {
            OracleUtils.ReportInfo memory reportInfo;
            SetPricesInnerCache memory innerCache;
```

Reply:
I meant that the loops in solidity are as it follows:

```
for (uint256 i; i < params.tokens.length; i++) {
    for (uint256 j = 0; j < signers.length; j++) {
    }
    for (uint256 j = 0; j < signers.length; j++) {
    }
}
```

But it looks like you're doing something like:
```
for (uint256 i; i < signers.length; i++) {
    for (uint256 j; j < signers.length; j++) {
    }
}
```

Reply:
Apologies for the oversight. The latest commit should resolve the issue.